internal/pkg/vm/internal/vmcontext/actor_state_handle.go

```diff
@@ -9,62 +9,110 @@ import (
+	readonlyHead := *h.head
 
-	// Dragons: cleanup after changing `Get` to `Get(cid, interface{})`
-	if err := encoding.Decode(rawstate, obj); err != nil {
-		runtime.Abort("Could not deserialize state")
+	if readonlyHead == cid.Undef {
+		runtime.Abort("Nil state can not be accessed via Readonly(), use Transaction() instead")
 	}
 
-	// clear head value so that it cannot be taken again
-	h.head = nil
+	h.get(readonlyHead, obj)
+
+	// add validation
+	h.addCidValidation(obj, readonlyHead)
 }
 
-// UpdateRelease updates the actor state and releases the lock on it.
+type transactionFn = func() (interface{}, error)
+
+// Transaction loads a mutable version of the state into the `obj` argument and protects
+// the execution from side effects.
 //
-// No future calls to `Take` are allowed on this object.
-func (h *actorStateHandle) UpdateRelease(obj interface{}) {
-	// Review: for @spec why are all this checks here?
-	// rt._checkRunning()
-	// rt._checkActorStateAcquired()
-	if h.head != nil {
-		runtime.Abort("Must call Take() before calling UpdateRelease()")
-	}
+// The second argument is a function which allows the caller to mutate the state.
+//
+// The new state will be commited if there are no errors returned.
+// Note: if an error is returned, the state changes will be DISCARDED and the reference will revert back.
+//
+// WARNING: If the state is modified AFTER the function returns, the execution will Abort.
+//	        The state is mutable ONLY inside the lambda.
+//
+// Transaction can be thought of as having the following signature:
+//
+// `Transaction(F) -> (T, Error) where F: Fn(S) -> (T, error), S: ActorState`.
+//
+// Note: the actual Go signature is a bit different due to the lack of type system magic,
+//       and also wanting to avoid some unnecesary reflection.
+//
+// Review: we might want to spend an hour or four making the signature look like it's supposed to..
+// Hack: In order to know `S` and save some code, the actual signature looks like:
+//       `Transaction(S, F) where S: ActorState, F: Fn() -> (T, Error)`.
+//
+// # Usage
+//
+// ```go
+// var state SomeState
+// ret, err := ctx.StateHandke().Transaction(&state, func() (interface{}, error) {
+//   // make some changes
+//	 st.ImLoaded = True
+//   return st.Thing, nil
+// })
+// // state.ImLoaded = False // BAD!! state is readonly outside the lambda
+// ```
+func (h *actorStateHandle) Transaction(obj interface{}, f transactionFn) (interface{}, error) {
+	// load state from storage
+	oldcid := *h.head
+	h.get(oldcid, obj)
 
-	storage := h.ctx.Storage()
-	oldcid := storage.Head()
+	// call user code allowing mutation but not side-effects
+	h.ctx.AllowSideEffects(false)
+	out, err := f()
+	h.ctx.AllowSideEffects(true)
+
+	// forward user error
+	if err != nil {
+		// re-load state from storage
+		// Note: this will throw away any changes done to the object
+		h.get(oldcid, obj)
```

Comment:
I don't think this is correct. Prior to this function call, obj was some arbitrary object pointer, probably nil. This does not restore it to its old state, but to a different state (equiv to Readonly). A correct rollback would be something like a deepcopy, but I don't think it's worth it. I propose instead that the obj will be left as the transaction left it, but won't be written to storage or have the head updated.